File: backend/api/app/services/audit_service.py
from sqlalchemy.orm import Session

from app.models.audit_log import AuditLog


def log_action(db: Session, action: str, user_id: int | None = None, document_id: int | None = None,
               details: dict | None = None, ip_address: str | None = None) -> None:
    entry = AuditLog(
        action=action,
        user_id=user_id,
        document_id=document_id,
        details=details,
        ip_address=ip_address,
    )
    db.add(entry)
    db.commit()

File: backend/src/services/audit.service.js
const prisma = require("../config/database");
const logger = require("../utils/logger");

async function logAction(action, userId, documentId, details, ipAddress) {
  try {
    await prisma.auditLog.create({
      data: { action, userId, documentId, details: details || {}, ipAddress },
    });
  } catch (err) {
    logger.error(`Failed to log audit: ${action}`, { error: err.message });
  }
}

module.exports = { logAction };
